### PR TITLE
Comm+colab service leave session

### PR DESCRIPTION
**Pull request by DesmondTo**
> 1. Leave session when one of the clients click `leave session` button.

https://user-images.githubusercontent.com/70432640/197017410-3af1a3df-616a-4141-b3be-de785df5da13.mov

> 2. Leave session when one of the clients close browser/disconnected.

https://user-images.githubusercontent.com/70432640/196989940-26cf3e1d-bad7-44dc-bf8d-3a8f977b59bc.mov

### CODE REVIEW

File: collaboration-service/index.js [hunk 1]
```diff
@@ -25,11 +25,34 @@ io.on('connection', (socket) => {
   socket.on('code-changed', (roomId, code) => {
     socket.to(roomId).emit('update-code', code);
   });
+
+  // leave-session: when user clicks on "leave session" button
+  // disconnecting: when user closes tab/disconnects
```
Comment (hsiaojietng) [inline collaboration-service/index.js:30]: Thanks for creating these legends to differentiate the actions!

File: communication-service/index.js [hunk 2]
```diff
@@ -28,6 +30,27 @@ io.on('connection', (socket) => {
...
+
+  // leave-session: when user clicks on "leave session" button
+  // disconnecting: when user closes tab/disconnects
+  const sessionEndEvents = ['leave-session', 'disconnecting'];
+
+  for (const event of sessionEndEvents) {
+    socket.on('leave-session', () => {
+      socket.rooms.forEach((room) => {
+        if (room !== socket.id) {
+          // Emit to other sockets in the same room this socket had joined
```
Comment (hsiaojietng) [inline communication-service/index.js:42]: This piece of code seems similar to the one in collaboration-service. If we have time, we can abstract it. But for now, this works!

**Review by hsiaojietng — APPROVED** [review]
LGTM, just some small nits.

File: communication-service/index.js (hunk 2 above)
Comment (DesmondTo) [reply communication-service/index.js:42]: I think it's ok to be the same since they are doing the same thing?

Comment (DesmondTo) [reply communication-service/index.js:44]: But yea will abstract these socket event listener out in future PR, thanks for pointing out.

Comment (hsiaojietng) [reply communication-service/index.js:44]: Alright sure!

**Pull request merged**